dpctl/dparray.py

```diff
@@ -0,0 +1,261 @@
+            # also check for array interface
+            new_obj = np.ndarray.__new__(
+                subtype,
+                shape,
+                dtype=dtype,
+                buffer=buffer,
+                offset=offset,
+                strides=strides,
+                order=order,
+            )
+            if hasattr(new_obj, array_interface_property):
+                dprint("buffer None new_obj already has sycl_usm")
+            else:
+                dprint("buffer None new_obj will add sycl_usm")
+                setattr(new_obj, array_interface_property, {})
+            return new_obj
+        else:
+            dprint("dparray::ndarray __new__ buffer not None and not sycl_usm")
+            nelems = np.prod(shape)
+            # must copy
+            ar = np.ndarray(
+                shape,
+                dtype=dtype,
+                buffer=buffer,
+                offset=offset,
+                strides=strides,
+                order=order,
+            )
+            nbytes = int(ar.nbytes)
+            buf = MemoryUSMShared(nbytes)
+            new_obj = np.ndarray.__new__(
+                subtype,
+                shape,
+                dtype=dtype,
+                buffer=buf,
+                offset=0,
+                strides=strides,
+                order=order,
+            )
+            np.copyto(new_obj, ar, casting="no")
+            if hasattr(new_obj, array_interface_property):
+                dprint("buffer None new_obj already has sycl_usm")
+            else:
+                dprint("buffer None new_obj will add sycl_usm")
+                setattr(new_obj, array_interface_property, {})
+            return new_obj
+
+    def __array_finalize__(self, obj):
+        dprint("__array_finalize__:", obj, hex(id(obj)), type(obj))
+        # When called from the explicit constructor, obj is None
+        if obj is None:
+            return
+        # When called in new-from-template, `obj` is another instance of our own
+        # subclass, that we might use to update the new `self` instance.
+        # However, when called from view casting, `obj` can be an instance of any
+        # subclass of ndarray, including our own.
+        if hasattr(obj, array_interface_property):
+            return
+        if isinstance(obj, np.ndarray):
+            ob = self
+            while isinstance(ob, np.ndarray):
+                if hasattr(ob, array_interface_property):
+                    return
+                ob = ob.base
+
+        # Just raise an exception since __array_ufunc__ makes all reasonable cases not
+        # need the code below.
+        raise ValueError(
+            "Non-USM allocated ndarray can not viewed as a USM-allocated one without a copy"
+        )
+
+    # Tell Numba to not treat this type just like a NumPy ndarray but to propagate its type.
+    # This way it will use the custom dparray allocator.
+    __numba_no_subtype_ndarray__ = True
```

Comment:
@DrTodd13 Do we really need this extra attribute? Can we not use the facts that the array is a subclass of NumPy.ndarray and it defines the `__sycl_usm_array_interface__` to recognize this class inside Numba?

Reply:
Numba is currently made to treat Numpy.ndarray subclasses as if they were Numpy.ndarrays.  Therefore, we can't just check for whether it is a ndarray subclass to keep those types separate without breaking backwards compatibility.

Also, you might want to keep types separate even if you don't need a special allocator.  So, for generality, we shouldn't specialize on something allocator or USM-specific inside Numba.